### PR TITLE
Exception hierarchy and handling updates

### PR DESCRIPTION
**Pull request by kiftio**
#### Exception Hierarchy Changes

- Add an `isRecoverable` attribute to CheckoutException and subclasses.
- Changes to CheckoutException hierarchy
  - `CheckoutSdkError` is renamed to `CheckoutSheetKitException` for consistency with other exception types and library naming
  - `AuthenticationException` is added - capture issues related to authentication / required login
  - `ConfigurationException` is added to capture any issues related to configuration. _Note: The more specific `CheckoutLiquidNotMigratedException` has been removed and will now be returned as a `ConfigurationException`_
  - `ClientException` and `HttpException` subclasses of `CheckoutUnavailable` have been added. HttpException exposes `statusCode`, which can be used to help make decisions about how to react or recover from an error.

#### Other

The equivalent of a `CheckoutLiquidNotMigratedException` is now only returned if we find an appropriate header from the backend, instead of using status code alone, which was too general.

When parsing errors from the bridge, we'll now differentiate between configuration errors

### Before you deploy

- [x] I have added tests to support my implementation
- [x] I have read and agree with the [contributing documentation](https://github.com/Shopify/checkout-sheet-kit-android/blob/main/.github/CONTRIBUTING.md)
- [x] I have read and agree with the [code of conduct documentation](https://github.com/Shopify/checkout-sheet-kit-android/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] I have updated the [README](https://github.com/Shopify/checkout-sheet-kit-android/blob/main/README.md) (if applicable).

### CODE REVIEW

File: lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutBridge.kt [hunk 1]
```diff
@@ -125,22 +128,26 @@ internal class CheckoutBridge(
...
-            CheckoutErrorGroup.EXPIRED -> CheckoutExpiredException(decodedError.reason)
-            else -> null
-        }
-        sheetKitError?.let {
-            eventProcessor.onCheckoutViewFailedWithError(sheetKitError)
+            CheckoutErrorGroup.UNRECOVERABLE -> eventProcessor.onCheckoutViewFailedWithError(
+                error = ClientException(decodedError.reason), isRecoverable = false,
+            )
+            CheckoutErrorGroup.EXPIRED -> eventProcessor.onCheckoutViewFailedWithError(
+                error = CheckoutExpiredException(decodedError.reason), isRecoverable = false,
```
Comment (kiftio) [inline lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutBridge.kt:146]: Not recoverable without first getting a new checkout URL / cart

File: lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt [hunk 2]
```diff
@@ -247,19 +247,32 @@ internal class CheckoutWebView(context: Context, attributeSet: AttributeSet? = n
             }
         }
 
+        private fun isRecoverable(statusCode: Int): Boolean {
+            return when (statusCode) {
+                TOO_MANY_REQUESTS -> true
```
Comment (kiftio) [inline lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt:252]: Feels worth a retry if we're rate limited. Although ideally we'd back off first

File: lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt [hunk 3]
```diff
@@ -247,19 +247,32 @@ internal class CheckoutWebView(context: Context, attributeSet: AttributeSet? = n
...
+                in CLIENT_ERROR -> false
```
Comment (kiftio) [inline lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt:253]: Probably worth checking this one.. Maybe client errors could also be resolved in an external view

File: lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutBridge.kt [hunk 4]
```diff
@@ -125,22 +128,26 @@ internal class CheckoutBridge(
...
+                error = ClientException(decodedError.reason ?: "Storefront was not configured properly."), isRecoverable = false,
             )
-            CheckoutErrorGroup.UNRECOVERABLE -> CheckoutUnavailableException(decodedError.reason)
-            CheckoutErrorGroup.EXPIRED -> CheckoutExpiredException(decodedError.reason)
-            else -> null
-        }
-        sheetKitError?.let {
-            eventProcessor.onCheckoutViewFailedWithError(sheetKitError)
+            CheckoutErrorGroup.UNRECOVERABLE -> eventProcessor.onCheckoutViewFailedWithError(
+                error = ClientException(decodedError.reason), isRecoverable = false,
```
[the remark this answers is not in the document]
Comment (markmur) [reply lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutBridge.kt:143]: This one is recoverable, no? Client side has crashed so we should trigger the fallback

File: lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt [hunk 5]
```diff
@@ -247,19 +247,32 @@ internal class CheckoutWebView(context: Context, attributeSet: AttributeSet? = n
...
-                val exception = when (errorCode) {
-                    HTTP_FORBIDDEN -> CheckoutUnavailableException("Forbidden")
-                    HTTP_NOT_FOUND -> CheckoutLiquidNotMigratedException()
-                    HTTP_GONE -> CheckoutExpiredException()
-                    else -> CheckoutUnavailableException(errorDescription)
+                val processor = checkoutBridge.getEventProcessor()
+                when (errorCode) {
+                    HTTP_NOT_FOUND -> processor.onCheckoutViewFailedWithError(
+                        CheckoutLiquidNotMigratedException(), false
+                    )
```
[the remark this answers is not in the document]
Comment (kiftio) [reply lib/src/main/java/com/shopify/checkoutsheetkit/CheckoutWebView.kt:268]: 👍 was doing this now

**Pull request merged**